src/main/java/org/broadinstitute/hellbender/engine/LocusWalker.java

```diff
@@ -0,0 +1,115 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMReadGroupRecord;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.downsampling.DownsamplingMethod;
+import org.broadinstitute.hellbender.utils.locusiterator.LocusIteratorByState;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * A LocusWalker is a tool that processes reads that overlaps a single position in a reference at a time from
+ * one or multiple sources of reads, with optional contextual information from a reference and/or sets of
+ * variants/Features.
+ *
+ * LocusWalker authors must implement the apply() method to process each position, and may optionally implement
+ * onTraversalStart() and/or onTraversalDone().
+ *
+ * @author Daniel Gómez-Sánchez (magicDGS)
+ */
+public abstract class LocusWalker extends GATKTool {
+
+	/**
+	 * LocusWalkers requires read sources
+	 */
+	@Override
+	public boolean requiresReads() {
+		return true;
+	}
+
+	/**
+	 * Does this tool require deletions in the AlignmentContext? Tools that do should override to return true.
+	 *
+	 * @return {@code true} if this tool requires deletions, {@code false} otherwise
+	 */
+	public boolean includeDeletions() {
+		return true;
+	}
+
+	/**
+	 * Get the information about how to downsample the reads. By default {@link org.broadinstitute.hellbender.utils.downsampling.DownsamplingMethod#NONE} is returned.
+	 * Subclasses should override it to provide own downsampling methods.
+	 *
+	 * @return the downsampling method for the reads
+	 */
+	public DownsamplingMethod getDownsamplingMethod() {
+		// TODO: change when downsampling command line options are implemented
+		return DownsamplingMethod.NONE;
+	}
+
+	/**
+	 * Marked final so that tool authors don't override it. Tool authors should override onTraversalStart() instead.
+	 */
+	@Override
+	protected final void onStartup() {
+		// Overridden only to make final so that concrete tool implementations don't override
+		super.onStartup();
```

Comment:
I think we should be able to get `-L` intervals working for this class without too much trouble. `intervalsForTraversal` *is* guaranteed to be sorted, so you can take advantage of that fact. If you set them on the reads data source here in `onStartup()`:

         if ( hasIntervals() ) {
            reads.setIntervalsForTraversal(intervalsForTraversal);
         }

and then wrap your `LocusIteratorByState` within another iterator that takes a sorted list of intervals + a nested iterator, you should be able to filter the `AlignmentContexts` returned to only those that overlap the user's intervals without any inefficiency.